app/forms/questionnaire_form.py

```diff
@@ -247,33 +295,36 @@ def _get_period_range_for_single_date(self, date_from, date_to):
         )
 
         handler = DateHandler(date_from, value_source_resolver)
-        from_min_period_date = handler.get_date_value("minimum")
-        from_max_period_date = handler.get_date_value("maximum")
+        from_min_period_date: datetime = handler.get_date_value("minimum")  # type: ignore # expect not None
+        from_max_period_date: datetime = handler.get_date_value("maximum")  # type: ignore # expect not None
 
         handler.answer_schema = date_to
-        to_min_period_date = handler.get_date_value("minimum")
-        to_max_period_date = handler.get_date_value("maximum")
+        to_min_period_date: datetime = handler.get_date_value("minimum")  # type: ignore # expect not None
+        to_max_period_date: datetime = handler.get_date_value("maximum")  # type: ignore # expect not None
```

Comment:
For things like this, this ignore is only needed because we are reusing a function that could return None or a datetime but we know in this instance they should exist, therefore you could refactor the existing method to always return a date.

```py
    def get_date_value_if_exists(self, key: str) -> Optional[datetime]:
        return self.get_date_value(key) if key in self.answer_schema else None

    def get_date_value(self, key:str) -> datetime:
        """
        Gets attributes within a minimum or maximum of a date field and validates that the entered date
        is valid.
        """
        date_value = self.get_referenced_date(key)

        if "offset_by" in self.answer_schema[key]:
            offset = self.answer_schema[key]["offset_by"]
            date_value = self.transform_date_by_offset(date_value, offset)

        return date_value
```
So we use `get_date_value_if_exists` when None is acceptable and the other when we expect to return a value. 

Thoughts?

Reply:
Sure, I was just avoiding it for the sake of pleasing a type hint

Reply:
I was thinking we type ignore only minor things that say can be `Any` etc but if it can technically break at runtime, it is worth handling if it is only a minor change.

Reply:
The values possibly returned by all the functions called in this function (`get_referenced_date`, `transform_date_by_offset`) are all optional too, so `get_date_value` as you've refactored here is still have an optional return type. This little gathering of code has this unusual pattern of having an optional main argument which, if it's falsy, returns `None`. I'd normally expect to see callers of the functions do the checking of values before calling the function, which led to [this](https://github.com/ONSdigital/eq-questionnaire-runner/pull/665#discussion_r716990564) change, which did break stuff. I'm happy to delve deeper and refactor these but for the sake of this single type hint, is it worth it?

Reply:
The typing for `get_referenced_date` and `transform_date_by_offset` implies they both always return a datetime, so when would it be optional?. Have I misunderstood or is the typing not correct right now?.

Reply:
`get_referenced_date` returns: `return convert_to_datetime(value) if isinstance(value, str) else None`
`convert_to_datetime` returns None if the arg is falsy
`transform_date_by_offset` returns the original arg if it's falsy

Reply:
Re: prior to your change, this is one of those where we shouldn't get `None` as validator should guard against it, but runner doesn't know that and technically should guard itself. We could simply raise an exception here if we get back `None` instead of a `datetime`, then the typing would behave as expected and can stay as `Optional[datetime]` and code covers NoneTypes. so that would be remove

Reply:
All done -raising an exception in the case the return type was None for either value.